pmd-java/src/main/java/net/sourceforge/pmd/lang/java/ast/AbstractJavaNode.java

```diff
@@ -32,6 +33,18 @@ public Scope getScope() {
         return scope;
     }
 
+    @Override
+    public final <R, P> R acceptVisitor(AstVisitor<? super P, ? extends R> visitor, P data) {
+        if (visitor instanceof JavaVisitor) {
+            return this.acceptVisitor((JavaVisitor<? super P, ? extends R>) visitor, data);
+        }
+        return visitor.visitNode(this, data);
```

Comment:
Should we allow this or rather prevent this? trying to visit an AST with an incompatible visitor... I would argue to restrict the feature set at the beginning rather than allowing too much....

Reply:
I think yes, we can restrict it for the time being. Maybe in the future it will be clearer what to do in this case